lib/surface/compiler/ast_translator.ex

```diff
@@ -0,0 +1,281 @@
+defmodule Surface.Compiler.AstTranslator do
+  @behaviour Surface.Compiler.NodeTranslator
+
+  @constructs %{
+    "if" => Surface.Construct.If,
+    "for" => Surface.Construct.For,
+    "template" => Surface.Construct.Template,
+    "slot" => Surface.Construct.Slot
+  }
+
+  @directives [
+    Surface.Directive.ComponentProps,
+    Surface.Directive.Let,
+    Surface.Directive.Hook,
+    Surface.Directive.Values,
+    Surface.Directive.TagAttrs,
+    Surface.Directive.Events,
+    Surface.Directive.Show,
+    Surface.Directive.If,
+    Surface.Directive.For,
+    Surface.Directive.Debug
+  ]
+
+  alias Surface.AST
+  alias Surface.Compiler.Helpers
+  alias Surface.Compiler.Parser
+
+  defguardp is_component(context) when context.type in [AST.Component, AST.MacroComponent]
+  defguardp is_construct(context) when context.type == Surface.Construct
+  defguardp is_subblock(context) when context.type == Surface.Construct.SubBlock
+  defguardp is_tag(context) when context.type in [AST.Tag, AST.VoidTag]
+
+  def handle_comment(state, _comment), do: {state, :ignore}
+
+  def handle_literal(state, value) do
+    {state, %AST.Literal{value: value}}
+  end
+
+  def handle_end(_state, children) do
+    children
+  end
+
+  def handle_interpolation(state, expression, parse_meta) do
+    ast_meta = to_meta(state, parse_meta)
+
+    {state,
+     %AST.Interpolation{
+       original: expression,
+       value: Helpers.interpolation_to_quoted!(expression, ast_meta),
+       meta: ast_meta
+     }}
+  end
+
+  def handle_attribute(state, context, name, value, attr_meta) do
+    meta = to_meta(state, attr_meta)
+
+    normalized_name = Surface.Directive.normalize_name(name)
+
+    directive =
+      Enum.find(@directives, fn directive ->
+        directive.matches?(context.type, normalized_name)
+      end)
+
+    if directive do
+      {String.to_atom(normalized_name),
+       %Surface.Directive{
+         module: directive,
+         original_name: name,
+         name: normalized_name,
+         value: parse_value(state, directive.type(), name, value, meta),
+         meta: meta
+       }}
```

Comment:
Is there a reason not to create an `handle_directive` and move the condition into the parser?

Reply:
though that would be a bit simpler, the parser doesn't quite have enough knowledge to tell if something's a directive or just an attribute